MetadataExtractor/IO/RandomAccessStream.cs

```diff
@@ -0,0 +1,561 @@
+            {
+                if ((index + bytesRequested) < p_streamLength)
+                    return bytesRequested;
+                else if (index > p_streamLength)
+                    return 0;
+                else
+                    return p_streamLength - index;
+            }
+
+
+            var endIndex = index + bytesRequested - 1;
+            if (endIndex < 0) endIndex = 0;
+
+            // Maybe don't check this?
+            if (endIndex > int.MaxValue)
+                return 0;
+
+            // zero-based
+            long chunkstart = index / p_chunkLength;
+            long chunkend = ((index + bytesRequested) / p_chunkLength) + 1;
+
+
+            if (!p_chunks.ContainsKey(chunkstart))
+            {
+                if (!CanSeek)
+                    chunkstart = p_chunks.Count == 0 ? 0 : p_chunks.Keys.Max() + 1;
+            }
+
+            for (var i = chunkstart; i < chunkend; i++)
+            {
+                if (!p_chunks.ContainsKey(i))
+                {
+                    p_isStreamFinished = false;
+
+                    // chunkstart can be anywhere. Try to seek
+                    if (CanSeek)
+                        p_inputStream.Seek(i * p_chunkLength, SeekOrigin.Begin);
+
+                    byte[] chunk = new byte[p_chunkLength];
+
+                    var totalBytesRead = 0;
+                    while (!p_isStreamFinished && totalBytesRead != p_chunkLength)
+                    {
+                        var bytesRead = p_inputStream.Read(chunk, totalBytesRead, p_chunkLength - totalBytesRead);
+
+                        if (bytesRead == 0)
+                        {
+                            // the stream has ended, which may be ok
+                            p_isStreamFinished = true;
+                            p_streamLength = i * p_chunkLength + totalBytesRead;
+
+                            // check we have enough bytes for the requested index
+                            if (endIndex >= p_streamLength)
+                            {
+                                TotalBytesRead += totalBytesRead;
+                                p_chunks.Add(i, chunk);
+                                return (index + bytesRequested) <= p_streamLength ? bytesRequested : p_streamLength - index;
+                            }
+                        }
+                        else
+                        {
+                            totalBytesRead += bytesRead;
+                        }
+                    }
+
+                    TotalBytesRead += totalBytesRead;
+                    p_chunks.Add(i, chunk);
+                }
+            }
+
+            if (p_isStreamFinished)
+                return (index + bytesRequested) <= p_streamLength ? bytesRequested : 0;
+            else
+                return bytesRequested;
+        }
+
+        /// <summary>
+        /// Records the total bytes buffered
+        /// </summary>
+        public long TotalBytesRead { get; private set; } = 0;
```

Comment:
Is this useful, other than for debugging?

Reply:
No, just debugging

Reply:
Perhaps wrap it and usages with `#if DEBUG`/`#endif`.

Reply:
Wrapped in the #if